### PR TITLE
CO-1995 Changed welcome active letter generation

### PR DESCRIPTION
**Pull request by quentingigon**

### CODE REVIEW

File: partner_communication_switzerland/migrations/10.0.2.4.6/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+def migrate(env, version):
+    if not version:
+        return
+
+    sponsorships = env['recurring.contract'].search([
+        ('activation_date', '!=', None)
+    ])
+    for sponsorship in sponsorships:
+        # if activation date is more than 24h old
+        if sponsorship.activation_date <= fields.Datetime.to_string(
```
Comment (ecino) [inline partner_communication_switzerland/migrations/10.0.2.4.6/post-migration.py]: Here you compare strings. It's safer to compare date objects to determine which is older. You can use `fields.Datetime.from_string(sponsorship.activation_date)`

File: partner_communication_switzerland/models/contracts.py [hunk 2]
```diff
@@ -417,7 +427,8 @@ def contract_waiting(self):
                       s.correspondent_id.email and s.partner_id.ref !=
                       '1502623')
         welcome.write({
-            'sds_state': 'waiting_welcome'
+            'sds_state': 'waiting_welcome',
+            'start_date': datetime.today()
```
Comment (ecino) [inline partner_communication_switzerland/models/contracts.py]: This should be done in module `recurring_contract` instead.

File: sponsorship_switzerland/models/contracts.py [hunk 3]
```diff
@@ -213,7 +213,10 @@ def contract_active(self):
 
     @api.multi
     def contract_waiting_mandate(self):
-        self.write({'state': 'mandate'})
+        self.write({
+            'state': 'mandate',
+            'start_date': datetime.today()
```
Comment (ecino) [inline sponsorship_switzerland/models/contracts.py]: This should be done in module `sponsorship_switzerland` instead.

File: partner_communication_switzerland/migrations/10.0.2.4.6/post-migration.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+from datetime import datetime, timedelta
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate(use_env=True)
+def migrate(env, version):
+    if not version:
+        return
+
+    sponsorships = env['recurring.contract'].search([
```
Comment (ecino) [inline partner_communication_switzerland/migrations/10.0.2.4.6/post-migration.py]: You could replace this whole script by a SQL query which will be much faster :

1. Remove openupgrade.migrate decorator
2. `env` function parameter becomes `cr`
3.

```
cr.execute("""
update recurring_contract
set welcome_active_letter_sent=true
where activation_date <= %s
""", fields.Datetime.to_string(datetime.today() - timedelta(days=1)))
```

**Review by ecino — CHANGES_REQUESTED** [review]

File: sponsorship_switzerland/models/contracts.py (hunk 3 above)
Comment (quentingigon) [reply sponsorship_switzerland/models/contracts.py]: We are in module `sponsorship_switzerland`, no ?

**Comment by quentingigon** [comment]
So for the start_date, now I set it up in compassion-accounting/recurring_contract/recurring_contract in the write method

**Review by ecino — APPROVED** [review]

**Pull request merged**